### PR TITLE
Description of mptable.rs module

### PR DESCRIPTION
**Pull request by boardslayer**

### CODE REVIEW

**Review by tmibvishal — APPROVED** [review]

File: mptable.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# [mptable](mptable.rs)
```
Comment (tmibvishal) [inline mptable.md:1]: Testing 2

**Review by tmibvishal — CHANGES_REQUESTED** [review]

File: mptable.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
```
Comment (ArkaMandal2001) [inline mptable.md:2]: Testing

**Review by ArkaMandal2001 — COMMENTED** [review]
Testing

File: mptable.md [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+## Introduction
```
Comment (amar512-prog) [inline mptable.md:3]: Testing

File: mptable.md [hunk 4]
```diff
@@ -0,0 +1,142 @@
+# [mptable](mptable.rs)
+
+## Introduction 
+
+A multiprocessor table is one of the two ways defined in the [MultiProcessor Specification V1.4](https://pdos.csail.mit.edu/6.828/2008/readings/ia32/MPspec.pdf) through which an operating system can get information about the underlying multiprocessor configuration. Floating-pointer structure is the minimal for this where as Table is the maximal data structure representing this. 
+
```
Comment (Laxman824) [inline mptable.md:6]: Testing

File: mptable.md (hunk 2 above)
Comment (SwarajGawande) [inline mptable.md:2]: Testing

File: mptable.md [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+From Chapter 4 of the MP Specification: 
+> This table is optional. The table is composed of a base section and an extended section. The base section contains entries that are completely backwards compatible with previous versions of this specification. The extended section contains additional entry types. 
+> 
+> The MP configuration table contains explicit configuration information about APICs, processors, buses, and interrupts. The table consists of a header, followed by a number of entries of various types. The format and length of each entry depends on its type. When present, this configuration table must be stored either in a non-reported system RAM or within the BIOS read-only memory space.
```
Comment (jayantmishra99) [inline mptable.md:10]: Testing

**Review by jayantmishra99 — COMMENTED** [review]
reviewed

File: mptable.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+## MpTable in Rust
+Within the rust code for implementing this structure we have: 
+```rust
+pub struct MpTable {
+    irq_num: u8,
+    cpu_num: u8,
+}
+```
+`cpu_num` allows the configuring of the number of vCPUs. `irq_num` refers to the number of intrurpt requests allowed within the configuration.
```
Comment (codenet) [inline mptable.md]: the configuring -> configuring
intrurpt -> interrupt

File: mptable.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+```rust
+    let mut checksum: u8 = 0;
+    let max_ioapic_id = self.cpu_num + 1;
+```
+
+`checksum` is allocated once the mptable has been finalised. `max_ioapic_id` is defined as the maximum value the I/O APIC is allowed to take. It is later checked while setting up the kvm interrupt routing. 
+
+> I/O APICs contain a redirection table, which is used to route the interrupts it receives from peripheral buses to one or more local APICs. 
+
+A memory object is further initialised with the relevant parameters. A `MpfIntel` object is defined using prefefined values for `signature`, `length` and `specification`. Checksum for this allocation is calculated and stored. The object is written in the memory and the memory pointer moved to the next location. We note the starting point here with the following statements. A space is left in the middle for the header which will be added after the rest of the configuration.
```
Comment (codenet) [inline mptable.md]: prefefined

File: mptable.md [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+```rust
+let table_base = base_mp;
+base_mp = base_mp.unchecked_add(mem::size_of::<MpcTable>() as u64);
+```
+
+Base pointer for the table is now set where all the specification of the virtualised resources shall be saved. Each of the vCPUs when stored in the table are allocated with the following data as per the intel spec. This is the written into the current `base_mp` value and the value is moved to the next available space.
```
Comment (codenet) [inline mptable.md]: ~~the~~ written

File: mptable.md [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                mpspec::CPU_BOOTPROCESSOR as u8
+            } else {
+                0
+            },
+        cpufeature: CPU_STEPPING,  // CPU FLAGS: BP
+        featureflag: CPU_FEATURE_APIC | CPU_FEATURE_FPU, //FEATURE FLAGS 8
+        ..Default::default()
+});
+```
+The are defined as follows
```
Comment (codenet) [inline mptable.md]: These

**Review by codenet — APPROVED** [review]
Very nice

**Pull request closed without merging**